profiler/src/ProfilerEngine/Datadog.Profiler.Native/AdaptiveSampler.cpp

```diff
@@ -0,0 +1,194 @@
+#include "AdaptiveSampler.h"
+
+#include <algorithm>
+#include <stdexcept>
+
+void AdaptiveSampler::Counts::AddTest()
+{
+    _testCount.fetch_add(1);
+}
+
+bool AdaptiveSampler::Counts::AddSample(int64_t limit)
+{
+    int64_t previousValue;
+    int64_t newValue;
+
+    do
+    {
+        previousValue = _sampleCount.load();
+        newValue = std::min<int64_t>(previousValue + 1, limit);
+    } while (!_sampleCount.compare_exchange_strong(previousValue, newValue));
+
+    return newValue < limit;
+}
+
+void AdaptiveSampler::Counts::AddSample()
+{
+    _sampleCount.fetch_add(1);
+}
+
+void AdaptiveSampler::Counts::Reset()
+{
+    _testCount = 0;
+    _sampleCount = 0;
+}
+
+int64_t AdaptiveSampler::Counts::SampleCount()
+{
+    return _sampleCount;
+}
+
+int64_t AdaptiveSampler::Counts::TestCount()
+{
+    return _testCount;
+}
+
+AdaptiveSampler::AdaptiveSampler(
+    std::chrono::milliseconds windowDuration,
+    int32_t samplesPerWindow,
+    int32_t averageLookback,
+    int32_t budgetLookback,
+    std::function<void()> rollWindowCallback) :
+    _timer([this] { RollWindow(); }, windowDuration),
+    _totalCountRunningAverage(0),
+    _rollWindowCallback(std::move(rollWindowCallback)),
+    _avgSamples(0)
+{
+    if (averageLookback < 1)
```

Comment:
Would it be possible to have a default value and log an error instead of throwing an exception?

Reply:
Done